### PR TITLE
 Merge feature-NodeAccess and feature-CNS branch to dev

### PR DESCRIPTION
**Pull request by chaychen2005**

### CODE REVIEW

File: fisco-bcos/Fake.h [hunk 1]
```diff
@@ -134,7 +134,10 @@ class FakeBlockChain : public BlockChainInterface
         return CommitResult::OK;
     }
 
-    void setGroupMark(std::string const& groupMark) override {}
+    void checkAndBuildGenesisBlock(GenesisBlockParam const& initParam) override {}
+
+    dev::h512s minerList() override { return dev::h512s(); };
+    dev::h512s observerList() override { return dev::h512s(); };
```
Comment (bxq2011hust) [inline fisco-bcos/Fake.h]: Should `minerList` and `observerList` const function?

File: libstorage/MemoryTableFactory.cpp [hunk 2]
```diff
@@ -86,6 +88,7 @@ Table::Ptr MemoryTableFactory::openTable(const string& tableName)
 
     memoryTable->init(tableName);
     m_name2Table.insert({tableName, memoryTable});
+    STORAGE_LOG(TRACE) << "open " << tableName << " successfully.";
```
Comment (cyjseagull) [inline libstorage/MemoryTableFactory.cpp]: recommend to remove this log

Comment (chaychen2005) [reply libstorage/MemoryTableFactory.cpp]: Removed.

File: libinitializer/P2PInitializer.cpp [hunk 3]
```diff
@@ -81,6 +83,23 @@ void P2PInitializer::initConfig(boost::property_tree::ptree const& _pt)
             }
         }
 
+        std::vector<std::string> crl;
+        /// CRL means certificate rejected list, CRL optional in config.ini
```
[the remark this answers is not in the document]
Comment (chaychen2005) [reply libinitializer/P2PInitializer.cpp]: Done.

File: libblockchain/BlockChainImp.cpp [hunk 4]
```diff
@@ -195,21 +221,100 @@ void BlockChainImp::setGroupMark(std::string const& groupMark)
...
+            if (!node)
+                return list;
+
+            if ((node->getField(blockverifier::NODE_TYPE) == type) &&
+                (boost::lexical_cast<int>(node->getField(blockverifier::NODE_KEY_ENABLENUM)) <=
+                    blockNumber))
+            {
+                h512 nodeID = h512(node->getField(blockverifier::NODE_KEY_NODEID));
+                list.push_back(nodeID);
+                LOG(TRACE) << "Add nodeID [nodeID/idx]: " << toHex(nodeID) << "/" << i << std::endl;
```
[the remark this answers is not in the document]
Comment (chaychen2005) [reply libblockchain/BlockChainImp.cpp]: Deleted.

File: libblockchain/BlockChainImp.cpp [hunk 5]
```diff
@@ -195,21 +221,100 @@ void BlockChainImp::setGroupMark(std::string const& groupMark)
...
+            BLOCKCHAIN_LOG(INFO) << "[#setGroupMark] Already have the 0th block, 0th groupMark is "
+                                    "equal to file groupMark.";
         }
         else
         {
             BLOCKCHAIN_LOG(WARNING)
-                << "[#setGroupMark] Already have the 0th block, groupMark does not allow "
-                   "modification! [groupMark]: "
-                << "[" << asString(block->header().extraData(0)) << "]";
+                << "[#setGroupMark] Already have the 0th block, 0th groupMark:"
```
[the remark this answers is not in the document]
Comment (chaychen2005) [reply libblockchain/BlockChainImp.cpp]: Done.

File: libblockchain/BlockChainImp.cpp [hunk 6]
```diff
@@ -195,21 +221,100 @@ void BlockChainImp::setGroupMark(std::string const& groupMark)
...
     else
     {
-        if (groupMark.compare(asString(block->header().extraData(0))) == 0)
+        /// compare() return 0 means equal!
+        /// If not equal, only print warning, willnot kill process.
+        if (!initParam.groupMark.compare(asString(block->header().extraData(0))))
         {
-            BLOCKCHAIN_LOG(INFO) << "[#setGroupMark] Already have the 0th block [groupMark]: "
-                                 << "[" << asString(block->header().extraData(0)) << "]";
+            BLOCKCHAIN_LOG(INFO) << "[#setGroupMark] Already have the 0th block, 0th groupMark is "
```
[the remark this answers is not in the document]
Comment (chaychen2005) [reply libblockchain/BlockChainImp.cpp]: Done.

File: libconsensus/pbft/PBFTEngine.cpp [hunk 7]
```diff
@@ -1181,56 +1181,28 @@ void PBFTEngine::updateMinerList()
...
-                        << i;
-                }
-            }
-        }
         UpgradeGuard ul(l);
-        m_minerList = miner_list;
+        m_minerList = m_blockChain->minerList();
         /// to make sure the index of all miners are consistent
         std::sort(m_minerList.begin(), m_minerList.end());
         m_lastObtainMinerNum = m_highestBlock.number();
```
[the remark this answers is not in the document]
Comment (chaychen2005) [reply libconsensus/pbft/PBFTEngine.cpp]: Done.

File: libblockchain/BlockChainImp.cpp [hunk 8]
```diff
@@ -180,6 +182,30 @@ void BlockChainImp::setGroupMark(std::string const& groupMark)
             tb->insert(lexical_cast<std::string>(block->blockHeader().number()), entry);
         }
 
+        tb = mtb->openTable(SYS_MINERS);
```
[the remark this answers is not in the document]
Comment (chaychen2005) [reply libblockchain/BlockChainImp.cpp]: SYS_MINERS provides writing interfaces but not reading interfaces.
Updating minerList and observerList will be done in resetConfig function.

**Pull request merged**